src/mpi/request/testall.c

```diff
@@ -265,13 +159,126 @@ int MPI_Testall(int count, MPI_Request array_of_requests[], int *flag,
 #   endif /* HAVE_ERROR_CHECKING */
 
     /* ... body of routine ...  */
+    MPIR_CHKLMEM_DECL(1);
+
+    /* Convert MPI request handles to a request object pointers */
+    if (count > MPIR_REQUEST_PTR_ARRAY_SIZE)
+    {
+        MPIR_CHKLMEM_MALLOC_ORJUMP(request_ptrs, MPIR_Request **,
+                count * sizeof(MPIR_Request *), mpi_errno, "request pointers", MPL_MEM_OBJECT);
+    }
+
+    for (i = 0; i < count; i++)
+    {
+        if (array_of_requests[i] != MPI_REQUEST_NULL)
+        {
+            MPIR_Request_get_ptr(array_of_requests[i], request_ptrs[i]);
+            /* Validate object pointers if error checking is enabled */
+#           ifdef HAVE_ERROR_CHECKING
+            {
+                MPID_BEGIN_ERROR_CHECKS;
+                {
+                    MPIR_Request_valid_ptr( request_ptrs[i], mpi_errno );
+                    if (mpi_errno) goto fn_fail;
+                }
+                MPID_END_ERROR_CHECKS;
+            }
+#           endif
+        }
+        else
+        {
+            request_ptrs[i] = NULL;
+        }
+    }
 
-    mpi_errno = MPIR_Testall_impl(count, array_of_requests, flag, array_of_statuses);
+    mpi_errno = MPIR_Testall_impl(count, request_ptrs, flag, array_of_statuses);
     if (mpi_errno != MPI_SUCCESS) goto fn_fail;
+    
+    /* If one or more of the requests completes with an error, MPI_ERR_IN_STATUS is returned
+     * An error value will be present is elements of array_of_status associated with the requests. */
+    if (*flag == FALSE) {
+        for (i = 0; i < count; i++)
+        {
+            if (request_ptrs[i] != NULL)
+            {
+                if (MPIR_Request_is_complete(request_ptrs[i]))
+                {
+                    rc = MPIR_Request_get_error(request_ptrs[i]);
+                    if (rc != MPI_SUCCESS)
+                    {
+                        if (MPIX_ERR_PROC_FAILED == MPIR_ERR_GET_CLASS(rc) || MPIX_ERR_PROC_FAILED_PENDING == MPIR_ERR_GET_CLASS(rc))
+                            proc_failure = TRUE;
+                        mpi_errno = MPI_ERR_IN_STATUS;
+                    }
+                } else if (unlikely(MPIR_CVAR_ENABLE_FT &&
+                            MPID_Request_is_anysource(request_ptrs[i]) &&
+                            !MPID_Comm_AS_enabled(request_ptrs[i]->comm)))
+                {
+                    mpi_errno = MPI_ERR_IN_STATUS;
+                    MPIR_ERR_SET(rc, MPIX_ERR_PROC_FAILED_PENDING, "**failure_pending");
+                    status_ptr = (array_of_statuses != MPI_STATUSES_IGNORE) ? &array_of_statuses[i] : MPI_STATUS_IGNORE;
+                    if (status_ptr != MPI_STATUS_IGNORE) status_ptr->MPI_ERROR = rc;
+                    proc_failure = TRUE;
+                }
+            }
+        }
+    }
+
+    /* All requests have completed, or an error has occured, finalize or set errors. */
+    if (*flag == TRUE || mpi_errno == MPI_ERR_IN_STATUS)
```

Comment:
Since we are treating`flag` as a boolean I'd prefer `!*flag` or `*flag != FALSE` instead of `*flag == TRUE`.

Reply:
Yes I can change to `!= FALSE` -- I don't want to use just `*flag` because there is no guarantee that one will define `FALSE == 0` (just being paranoid here)